field/TinyLang/Field/Typed/TypeChecker.hs

```diff
@@ -186,71 +186,65 @@ withTypedBinOp R.BAt k = k T.BAt
-    => Uni f a -> R.Var -> m (T.UniVar f a)
-checkUniVar uni iden = do
-    Some uniVar@(T.UniVar varUni _) <- inferUniVar iden
-    let uniMismatch = typeMismatch uniVar uni varUni
-    withGeqUniM uni varUni uniMismatch uniVar
-
 {-| Type checking for expressions
 -}
-checkExpr
-    :: forall m f a. (MonadTypeChecker m, TextField f, KnownUni f a)
-    => R.Expr R.Var f -> m (T.Expr f a)
+checkExpr ::
+    forall m f a. (Monad m, TextField f, KnownUni f a)
+    => R.Expr R.Var f -> TypeCheckerT m (T.Expr f a)
 checkExpr (R.EIf l m n) = T.EIf <$> checkExpr l <*> checkExpr m <*> checkExpr n
 checkExpr m = do
     SomeOf mUni tM <- inferExpr m
     let uni = knownUni @f @a
     let uniMismatch = typeMismatch tM uni mUni
     withGeqUniM uni mUni uniMismatch tM
 
--- TODO:  Figure how to optimise
-checkProgram
-    :: forall m f. (MonadTypeChecker m, TextField f)
-    => R.Program R.Var f -> m (T.Program f)
-checkProgram (R.Program (R.Statements stmts)) =
-    T.Program . T.Statements <$> foldMapA checkStatement stmts
+checkProgram ::
+    forall m f. (Monad m, TextField f)
+    => R.Program R.Var f -> TypeCheckerT m (T.Program f)
+checkProgram (R.Program stmts) =
+    checkStatements stmts $ pure . T.Program
 
+checkStatements ::
+    forall m f. (Monad m , TextField f)
+    => R.Statements R.Var f -> forall r. (T.Statements f -> TypeCheckerT m r) -> TypeCheckerT m r
+checkStatements (R.Statements stmts) kont =
+    runContT (foldMapA (ContT . checkStatement) stmts) $ kont . T.Statements
 
 {-| Type checking judgement for statements of form
 -}
-checkStatement
-    :: forall m f. (MonadTypeChecker m, TextField f)
-    => R.Statement R.Var f -> m [T.Statement f]
-checkStatement (R.ELet var m) = do
+checkStatement ::
+    forall m f. (Monad m , TextField f)
+    => R.Statement R.Var f -> forall r. ([T.Statement f] -> TypeCheckerT m r) -> TypeCheckerT m r
+checkStatement (R.ELet var m) kont = do
     Some uniVar@(T.UniVar uni _) <- inferUniVar var
-    T.withKnownUni uni $ (:[]) . T.ELet uniVar <$> checkExpr m
-checkStatement (R.EAssert m) =
-    (:[]) . T.EAssert <$> checkExpr m
--- TODO:  Inject var into environment of stmts
-checkStatement (R.EFor var start end stmts) =
-    foldMapA iter [start .. end]
-    where iter i = do
-              tVar <- makeVar $ R.unVar var
-              let uVar = T.UniVar Field tVar
-              tStmts <- foldMapA checkStatement stmts
-              pure $ T.ELet uVar (T.EConst . fromIntegral $ i) : tStmts
+    tM <- T.withKnownUni uni $ checkExpr m
+    withVar var $ kont [T.ELet uniVar tM]
+checkStatement (R.EAssert m) kont = do
+    tM <- checkExpr m
+    kont [T.EAssert tM]
+checkStatement (R.EFor var start end stmts) kont = do
+    T.Statements tStmts <- withVar var $ checkStatements stmts pure
+    tVar <- makeVar $ R.unVar var
+    let uVar = T.UniVar Field tVar
+    let iter i ikont = ikont $ T.ELet uVar (T.EConst . fromIntegral $ i) : tStmts
+    runContT (foldMapA (ContT . iter) [start .. end]) kont
```

Comment:
> I will probably add it as a type checking mechanism. Ideally I want to avoid interleaving of normalisation and typechecking or evaluation and typechecking.

Yeah, it would be best to avoid interleaving (partial) evaluation and type checking.